libecole/tests/src/dynamics/test-primalsearch.cpp

```diff
@@ -0,0 +1,209 @@
+		std::tie(done, action_set) = dyn.step_dynamics(model, action);
+		REQUIRE_FALSE(done);
+	}
+
+	SECTION("Handle SCIP special values - SCIP_UNKNOWN") {
+		auto [done, action_set] = dyn.reset_dynamics(model);
+		REQUIRE(action_set.has_value());
+		auto const& var_ids = action_set.value();
+		REQUIRE(var_ids.size() > 0);
+		auto const& action = filled_action(action_set, SCIP_UNKNOWN);
+		std::tie(done, action_set) = dyn.step_dynamics(model, action);
+		REQUIRE_FALSE(done);
+	}
+
+	SECTION("Handle SCIP special values - SCIP_INVALID") {
+		auto [done, action_set] = dyn.reset_dynamics(model);
+		REQUIRE(action_set.has_value());
+		auto const& var_ids = action_set.value();
+		REQUIRE(var_ids.size() > 0);
+		auto const& action = filled_action(action_set, SCIP_INVALID);
+		std::tie(done, action_set) = dyn.step_dynamics(model, action);
+		REQUIRE_FALSE(done);
+	}
+
+	SECTION("Handle SCIP special values - infinity") {
+		auto [done, action_set] = dyn.reset_dynamics(model);
+		REQUIRE(action_set.has_value());
+		auto const& var_ids = action_set.value();
+		REQUIRE(var_ids.size() > 0);
+		auto const& action = filled_action(action_set, SCIPinfinity(model.get_scip_ptr()));
+		std::tie(done, action_set) = dyn.step_dynamics(model, action);
+		REQUIRE_FALSE(done);
+	}
+
+	SECTION("Handle SCIP special values - minus infinity") {
+		auto [done, action_set] = dyn.reset_dynamics(model);
+		REQUIRE(action_set.has_value());
+		auto const& var_ids = action_set.value();
+		REQUIRE(var_ids.size() > 0);
+		auto const& action = filled_action(action_set, -SCIPinfinity(model.get_scip_ptr()));
+		std::tie(done, action_set) = dyn.step_dynamics(model, action);
+		REQUIRE_FALSE(done);
+	}
+
+	SECTION("Solve instance") {
+		auto [done, action_set] = dyn.reset_dynamics(model);
+		while (!done) {
+			REQUIRE(action_set.has_value());
+			std::tie(done, action_set) = dyn.step_dynamics(model, {});
+		}
+		REQUIRE(model.is_solved());
+	}
+
+	SECTION("Throw on invalid variable id") {
+		auto [done, action_set] = dyn.reset_dynamics(model);
+		REQUIRE_FALSE(done);
+		REQUIRE(action_set.has_value());
+		std::map<std::size_t, SCIP_Real> const action = {{model.variables().size(), 0.0}};
+		REQUIRE_THROWS_AS(dyn.step_dynamics(model, action), std::exception);
+	}
+}
+
+TEST_CASE("PrimalSearchDynamics handles limits", "[dynamics]") {
+	auto dyn = dynamics::PrimalSearchDynamics{1};
+	auto model = get_model();
+
+	SECTION("Node limit") {
+		auto const node_limit = GENERATE(0, 1, 2);
+		model.set_param("limits/totalnodes", node_limit);
+	}
+
+	SECTION("Time limit") {
+		auto const time_limit = GENERATE(0, 1, 2);
+		model.set_param("limits/time", time_limit);
+	}
+
+	auto [done, action_set] = dyn.reset_dynamics(model);
+	while (!done) {
+		REQUIRE(action_set.has_value());
+		std::tie(done, action_set) = dyn.step_dynamics(model, {});
+	}
+}
+
+TEST_CASE("PrimalSearchDynamics extreme parameterizations", "[dynamics]") {
+
+	SECTION("Infinite trial loop") {
```

Comment:
What is the difference with line 113 `SECTION("Solve instance")`, the same empty action set seems to be given there but the loop is not infinite?

Reply:
Here the environment does an infinite loop at the root node, and the episode end is triggered by a time limit.